### PR TITLE
wasmtime: update to v2.0.2.

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

### CODE REVIEW

**Comment by PiotrSikora** [comment]
You need to regenerate Bazel rules (see discussion in https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/309).

**Comment by phlax** [comment]
im hitting this

```console
$ cargo-raze --generate-lockfile
Error: Failed to load lockfile: /tmp/.tmp7bTIAD/Cargo.lock

Caused by:
    I/O operation failed: couldn't open /tmp/.tmp7bTIAD/Cargo.lock: No such file or directory (os error 2)

```

**Comment by phlax** [comment]
ive cleaned my cache and tried again - same result

if i try to run `cargo update` (iiuc this creates the missing file) - i get

```console
$ cargo update
    Updating crates.io index
    Updating git repository `https://github.com/bytecodealliance/wasmtime`
error: failed to select a version for the requirement `wasmtime = "^2.0.2"`
candidate versions found which didn't match: 0.37.0, 0.36.0, 0.35.3, ...
location searched: crates.io index
required by package `wasmtime-c-api-bazel v2.0.2 (/home/phlax/dev/envoydev/proxy-wasm-cpp-host/bazel/cargo/wasmtime)`
```

which is wierd because wasmtime defo has 2.0.2 on crates.io

**Comment by PiotrSikora** [comment]
The error is not very helpful, but I believe that updating Rust to a more recent version (`rustup update`) should fix it.

**Comment by phlax** [comment]
afaict this will need rules_rust to be updated - i think its being updated in the cargo generated files - but the repo has not been updated

**Comment by PiotrSikora** [comment]
Which version of `cargo-raze` are you using? The diff looks a quite large for such a small update.

Also, could you split the licenses fix into separate PR? I'll merge it right away.

**Comment by phlax** [comment]
```console
$ cargo install cargo-raze --version 0.12.0

```

**Comment by phlax** [comment]
https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/320

**Comment by PiotrSikora** [comment]
You need `cargo install cargo-raze --version 0.14.1`.

**Comment by PiotrSikora** [comment]
You need to remove the duplicate `select` statements from Bazel rules for `errno` and `rustix` crates (see: https://github.com/proxy-wasm/proxy-wasm-cpp-host/commit/f8f1c0706f59e02441004bc0bb13a26e44c79874).

**Comment by phlax** [comment]
im not 100% that i changed the right things

File: bazel/cargo/wasmtime/remote/BUILD.errno-0.2.8.bazel [hunk 1]
```diff
@@ -55,7 +55,6 @@ rust_library(
     ] + selects.with_or({
         # cfg(unix)
         (
-            "@rules_rust//rust/platform:wasm32-wasi",
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.errno-0.2.8.bazel:58]: This shouldn't be removed (the conditions for duplicated `select` deps were supposed to be combined, not removed). Could you restore this line?

File: bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel [hunk 2]
```diff
@@ -74,6 +74,13 @@ cargo_build_script(
             "@rules_rust//rust/platform:powerpc-unknown-linux-gnu",
             "@rules_rust//rust/platform:s390x-unknown-linux-gnu",
             "@rules_rust//rust/platform:x86_64-linux-android",
+        ): [
+            "@wasmtime__linux_raw_sys__0_0_46//:linux_raw_sys",
+        ],
+        "//conditions:default": [],
+    }) + selects.with_or({
+        # cfg(all(not(rustix_use_libc), not(miri), target_os = "linux", any(target_arch = "x86", all(target_arch = "x86_64", target_pointer_width = "64"), all(target_endian = "little", any(target_arch = "arm", all(target_arch = "aarch64", target_pointer_width = "64"), target_arch = "powerpc64", target_arch = "riscv64", target_arch = "mips", target_arch = "mips64")))))
+        (
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel:83]: Those `select` statements should be combined (i.e. remove those 7 lines).

File: bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel [hunk 3]
```diff
@@ -159,6 +150,11 @@ rust_library(
             "@rules_rust//rust/platform:powerpc-unknown-linux-gnu",
             "@rules_rust//rust/platform:s390x-unknown-linux-gnu",
             "@rules_rust//rust/platform:x86_64-linux-android",
+        ): [],
+        "//conditions:default": [],
+    }) + selects.with_or({
+        # cfg(all(not(rustix_use_libc), not(miri), target_os = "linux", any(target_arch = "x86", all(target_arch = "x86_64", target_pointer_width = "64"), all(target_endian = "little", any(target_arch = "arm", all(target_arch = "aarch64", target_pointer_width = "64"), target_arch = "powerpc64", target_arch = "riscv64", target_arch = "mips", target_arch = "mips64")))))
+        (
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel:157]: Those 5 lines should be removed.

File: bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel [hunk 4]
```diff
@@ -104,15 +104,6 @@ cargo_build_script(
...
         "//conditions:default": [],
-    }) + selects.with_or({
-        # cfg(windows)
-        (
-            "@rules_rust//rust/platform:i686-pc-windows-msvc",
-            "@rules_rust//rust/platform:x86_64-pc-windows-msvc",
-        ): [
-            "@wasmtime__windows_sys__0_36_1//:windows_sys",
-        ],
-        "//conditions:default": [],
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel:115]: Those lines should be restored.

File: bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel [hunk 5]
```diff
@@ -168,12 +164,8 @@ rust_library(
         ],
         "//conditions:default": [],
     }) + selects.with_or({
-        # cfg(all(not(rustix_use_libc), not(miri), target_os = "linux", any(target_arch = "x86", all(target_arch = "x86_64", target_pointer_width = "64"), all(target_endian = "little", any(target_arch = "arm", all(target_arch = "aarch64", target_pointer_width = "64"), target_arch = "powerpc64", target_arch = "riscv64", target_arch = "mips", target_arch = "mips64")))))
+        # cfg(any(rustix_use_libc, miri, not(all(target_os = "linux", any(target_arch = "x86", all(target_arch = "x86_64", target_pointer_width = "64"), all(target_endian = "little", any(target_arch = "arm", all(target_arch = "aarch64", target_pointer_width = "64"), target_arch = "powerpc64", target_arch = "riscv64", target_arch = "mips", target_arch = "mips64")))))))
         (
-            "@rules_rust//rust/platform:i686-unknown-linux-gnu",
-            "@rules_rust//rust/platform:x86_64-unknown-linux-gnu",
-            "@rules_rust//rust/platform:aarch64-unknown-linux-gnu",
-            "@rules_rust//rust/platform:arm-unknown-linux-gnueabi",
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel:176]: Those 4 lines should be restored.

File: bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel [hunk 6]
```diff
@@ -191,17 +183,8 @@ rust_library(
...
             "@wasmtime__errno__0_2_8//:errno",
-        ],
-        "//conditions:default": [],
-    }) + selects.with_or({
-        # cfg(windows)
-        (
-            "@rules_rust//rust/platform:i686-pc-windows-msvc",
-            "@rules_rust//rust/platform:x86_64-pc-windows-msvc",
-        ): [
-            "@wasmtime__windows_sys__0_36_1//:windows_sys",
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel:204]: Those lines should be restored.

File: bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel [hunk 7]
```diff
@@ -202,6 +202,7 @@ rust_library(
             "@rules_rust//rust/platform:x86_64-pc-windows-msvc",
         ): [
             "@wasmtime__windows_sys__0_36_1//:windows_sys",
+            "@wasmtime__libc__0_2_137//:libc",
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel:205]: This should be moved to the block before (after `errno`). See the diff.

File: bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel [hunk 8]
```diff
@@ -168,8 +168,9 @@ rust_library(
         ],
         "//conditions:default": [],
     }) + selects.with_or({
-        # cfg(all(not(rustix_use_libc), not(miri), target_os = "linux", any(target_arch = "x86", all(target_arch = "x86_64", target_pointer_width = "64"), all(target_endian = "little", any(target_arch = "arm", all(target_arch = "aarch64", target_pointer_width = "64"), target_arch = "powerpc64", target_arch = "riscv64", target_arch = "mips", target_arch = "mips64")))))
+        # cfg(any(rustix_use_libc, miri, not(all(target_os = "linux", any(target_arch = "x86", all(target_arch = "x86_64", target_pointer_width = "64"), all(target_endian = "little", any(target_arch = "arm", all(target_arch = "aarch64", target_pointer_width = "64"), target_arch = "powerpc64", target_arch = "riscv64", target_arch = "mips", target_arch = "mips64")))))))
         (
+
```
Comment (PiotrSikora) [inline bazel/cargo/wasmtime/remote/BUILD.rustix-0.35.13.bazel:173]: Nit: empty line (hopefully this doesn't trip the buildifier check).

**Pull request merged**